Speedup filter groups

GregorCH opened the pull request:
speed up IPET evaluation considerably through

- grouping all active filters and reusing
  already computed results from equal
  filters across filter groups
- reusing the same set of intersection rows
  across all filter groups
- replacing some apply operations by bitwise ands

fschloesser commented on ipet/evaluation/IPETFilter.py:
```diff
@@ -518,6 +569,33 @@ def getName(self):
+        dfindex = df.set_index(index).index
+
+        groups = df.groupby(index)
+        instancecount = groups.apply(len).max()
```
Is this always the number of instances? Or is this the maximal number of different instances? What happens if the different sets of instances are not subsets of each other?

GregorCH replied:
`instancecount` is the maximum frequency, how often *one* instance occurs. This is nice for concatening a run that is split over two test runs on a disjoint set of instances. Your second question is, I guess, what happens if two testruns have been run on disjoint subgroups, with different settings. I guess that this will produce useless tables, but the comparison is also meaningless. :)

fschloesser replied:
Ah okay, so this is after the reduction, and counts how often the instance spread over the columns?

fschloesser commented on test/FilterMethodsTest.py:
```diff
@@ -0,0 +1,83 @@
+                f1 = IPETFilter("A", "B", operator1)
+                f2 = IPETFilter("A", "B", operator2)
+                eq = f1.equals(f2)
+                self.assertEqual(eq, operator1 == operator2, "Bla")
```
Yay for testing, but may I suggest that we use a more meaningful errormessage?

GregorCH replied:
Yes, I will change this. Remind me to show you a best of of your meaningful error messages at some point in time.

GregorCH commented on ipet/evaluation/IPETEvalTable.py:
```diff
@@ -1305,8 +1305,8 @@ def getColumnFormatters(self, df):
 
         # display countercolumns as integer
         counting_columns = [dfcol for dfcol in df.columns if dfcol[l].startswith("_") and dfcol[l].endswith("_")]
```
I agree, but I did not even touch this line for this pull request.

fschloesser replied:
Yes, but you could have commented them out with the next two lines.
Why did you comment them out by the way?

GregorCH replied:
I reverted this. This slipped its way into this MR, sorry. "%d" cannot work with floating point NaN's. Nevermind. If this bug happens again, I can fix this in the beautify function of the Formatter class.

GregorCH commented on ipet/evaluation/IPETFilter.py:
```diff
@@ -461,6 +506,12 @@ class IPETFilterGroup(IpetNode):
 
+    # global data frame and index that are reused for several filter groups
+    _glbdf_ = None
+    _glbindex_ = None
```
There are actually two variables, and a comment above that explains them as well as I can.

fschloesser replied:
There is one additional one in the line following downstairs (l. 512)

GregorCH commented on ipet/evaluation/IPETFilter.py:
```diff
@@ -518,6 +569,33 @@ def getName(self):
+        IPETFilterGroup._glbinterrows_ = IPETFilterGroup.computeIntersectionRows(df, index)
+
+    @staticmethod
+    def computeIntersectionRows(df : pd.DataFrame, index):
```
It is the index of the subset of rows of the input frame considered as "intersection" for filter groups that have the "intersection" type (which is the default).

fschloesser replied:
I think a more detailed docstring would help here. If it is difficult to phrase maybe add a mini example.

fschloesser commented on ipet/evaluation/IPETFilter.py:
```diff
@@ -461,6 +506,12 @@ class IPETFilterGroup(IpetNode):
+    # global data frame and index that are reused for several filter groups
+    _glbdf_ = None
+    _glbindex_ = None
+    _glbinterrows_ = None
```
This is the third one in question. I apologize for not being able to use the comment functionality of github.

GregorCH replied:
I added a comment for this attribute.

GregorCH commented on ipet/evaluation/IPETEvalTable.py:
```diff
@@ -1733,5 +1740,36 @@ def aggregateTable(self, df):
+                    logging.debug("Reusing previously stored result for filter {}".format(f.getName()))
+                    fcol = stored_result
+                else:
+                    fcol = f.applyFilter(df, self.getRowIndex())
```
Very good point. Only for filters with index >= 1, a previous f should be considered.

fschloesser replied:
That makes sense. But as you do `applyFilter` here, shouldn't you also do this for the first one? (also otherwise the filter with index 1 does not have any result it could reuse)

GregorCH commented on ipet/evaluation/IPETEvalTable.py:
```diff
@@ -1733,5 +1740,36 @@ def aggregateTable(self, df):
+                # try to skip the computation of the filter column if an equal
+                # filter already has a stored result
+                #
+                stored_result = previous_f.getStoredResult(df)
```
This would lead to an additional level of "if" nesting.

fschloesser replied:
Ok

GregorCH commented:
Thanks for your review, I modified the corresponding places in the code, and added more documentation and meaningful assertion messages to the new test.

fschloesser commented:
Alright, this is ready to merge then.
